modules/internal/BytesStringCommon.chpl

```diff
@@ -273,14 +273,83 @@ module BytesStringCommon {
     }
   }
 
+  // Checks to see if r is inside the bounds of this and returns a finite
+  // range that can be used to iterate over a section of the string
+  //
+  // This function handles ranges of codepointIndex, byteIndex or numeric types.
+  // codepointIndex only makes sense for string, whereas the others can be used
+  // with both bytes and string
+  //
+  // If codepointIndex range was given, converts that to byte index range in the
+  // process.
+  proc getView(const ref x: ?t, r: range(?)) {
+    assertArgType(t, "getView");
+    if t == bytes && r.idxType == codepointIndex {
+      compilerError("codepointIndex ranges cannot be used with bytes in getView");
+    }
+
+    if t == bytes || r.idxType == byteIndex {
+      // cast the argument r to `int` to make sure that we are not dealing with
+      // byteIndex
+      const intR = r:range(int, r.boundedType, r.stridable);
+      if boundsChecking {
+        if !x.byteIndices.boundsCheck(intR) {
+          halt("range ", r, " out of bounds for " + t:string + " with ",
+               x.numBytes, " bytes");
+        }
+      }
+      return intR[x.byteIndices];
```

Comment:
I am curious. If one uses byte indexing to request a range of a string, can one create a string that is invalid UTF-8?

Reply:
Yes that can  happen with or without this PR. https://github.com/Cray/chapel-private/issues/937#issuecomment-622006746

We need to do something about it, but the path is not crystal clear to me at the moment.

Reply:
It seems like string slicing by byte indexes need to check for codepoint boundaries near the range start/end bytes. AFAIK this can even be done by looking at the first byte in the result and the last couple bytes in the result to check for bytes that are in the middle of UTF-8 sequences.

Reply:
I tend to agree, where I'd imagine the program would throw an error or halt if it wasn't well-aligned.

Reply:
(but in saying that, I don't mean to imply that this PR needs to address this case per se)

Reply:
I'll do that as a follow up.